Enable type storage for vector tags

loganharbour opened the pull request:
Closes #15282

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/525314/) on 5692327 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15284/site)

This comment will be updated on new commits.

loganharbour committed:
Enable type storage for vector tags

refs #15282

fdkong commented on framework/include/base/Assembly.h:
```diff
@@ -1506,6 +1511,14 @@ class Assembly
+   *
+   * Note that this will also clear the cache
+   */
+  void addCachedResidualInternal(NumericVector<Number> & residual, unsigned int tag_index);
```
`unsigned int` --> `TagID`

loganharbour replied:
But this isn't a `TagID`. It's an index into `_residual_vector_tags`, which contains only the tags that are residual tags. Therefore, if we have solution vector tags too, the indexing into `_residual_vector_tags` isn't the same as the ID of each tag in the vector.

fdkong commented on framework/include/base/Assembly.h:
```diff
@@ -1567,7 +1580,12 @@ class Assembly
+   * indexing into _sub_Re, _sub_Rn, _sub_Rl, _cached_residual_rows, and _cached_residual_values
+   */
+  unsigned int residualVectorTagIndex(const TagID tag_id) const;
+
```
Same here

fdkong commented on framework/include/problems/DisplacedProblem.h:
```diff
@@ -107,12 +107,16 @@ class DisplacedProblem : public SubProblem
+  virtual bool vectorTagExists(const TagID tag_id) const override;
+  virtual unsigned int
+  numVectorTags(const Moose::VectorTagType type = Moose::VECTOR_TAG_ANY) const override;
+  virtual const std::vector<VectorTag> &
```
Do we have to use `Moose::`?

loganharbour replied:
Yes.

fdkong commented on framework/include/utils/VectorTag.h:
```diff
@@ -0,0 +1,30 @@
+  bool operator==(const VectorTag & other) const;
+
+  /// The id associated with the vector tag
+  TagID _id;
```
We may add `TagID _local_id` as well to avoid querying later.

fdkong commented on framework/src/base/Assembly.C:
```diff
@@ -2975,79 +2975,90 @@ Assembly::setResidualBlock(NumericVector<Number> & residual,
 void
 Assembly::addResidualNeighbor(NumericVector<Number> & residual, TagID tag_id)
 {
+  auto & tag_Rn = _sub_Rn[residualVectorTagIndex(tag_id)];
```
`residualVectorTagIndex(tag_id)` should be repalced by "`_local_id`"

E.g.,

`_vector_tags[tag_id]._local_id.`

fdkong replied:
We may check if it is a residual tag as well

fdkong commented on framework/src/base/Assembly.C:
```diff
@@ -2975,79 +2975,90 @@ Assembly::setResidualBlock(NumericVector<Number> & residual,
+  {
+    mooseAssert(vector_tag._type == Moose::VECTOR_TAG_RESIDUAL, "Non-residual tag in residual");
+    if (_sys.hasVector(vector_tag._id))
+      addResidualNeighbor(_sys.getVector(vector_tag._id), vector_tag._id);
```
Or we just pass in `_local_id`

fdkong commented on framework/src/base/Assembly.C:
```diff
@@ -2233,21 +2234,20 @@ Assembly::init(const CouplingMatrix * cm)
   if (_block_diagonal_matrix && scalar_vars.size() != 0)
     _block_diagonal_matrix = false;
 
-  auto num_vector_tags = _subproblem.numVectorTags();
```
`auto num_vector_tags = _residual_vector_tags.size();`, and we do not need to change that many lines of code

fdkong commented on framework/src/base/Assembly.C:
```diff
@@ -4026,6 +4043,18 @@ Assembly::modifyFaceWeightsDueToXFEM(const Elem * elem, unsigned int side)
+{
+  for (unsigned int tag_index = 0; tag_index < _residual_vector_tags.size(); ++tag_index)
+    if (_residual_vector_tags[tag_index]._id == tag_id)
+      return tag_index;
```
This function will be called many times. It might be a good idea to let `VectorTag` store a `local id (tag index)`. If we do so, we can remove this function. It is a linear algorithm here.

fdkong commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -426,6 +426,13 @@ Coupleable::coupledVectorTagValue(const std::string & var_name, TagID tag, unsig
+               "in ",
+               _c_name,
+               ", but a vector tag with that ID does not exist");
+
```
`else`

fdkong commented on framework/src/problems/SubProblem.C:
```diff
@@ -67,49 +68,138 @@ SubProblem::SubProblem(const InputParameters & parameters)
+SubProblem::vectorTagExists(const TagName & tag_name) const
 {
-  auto tag_name_upper = MooseUtils::toUpper(tag_name);
+  mooseAssert(verifyVectorTags(), "Vector tag storage invalid");
```
Yes, I still like the original O(log(n)) algorithm that is better than O(n). This runtime will be used often.

fdkong replied:
We need  a map here

fdkong commented on framework/src/problems/SubProblem.C:
```diff
@@ -67,49 +68,138 @@ SubProblem::SubProblem(const InputParameters & parameters)
-               "specify a Transient Executioner.");
+  const auto tag_name_upper = MooseUtils::toUpper(tag_name);
+  for (const auto & vector_tag : _vector_tags)
+    if (vector_tag._name == tag_name_upper)
```
std::map

fdkong replied:
Or even better:`std::unordered_map`

fdkong commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -2201,19 +2201,19 @@ MooseVariableData<OutputType>::fetchDoFValues()
+  for (auto tag : active_coupleable_vector_tags)
+    if (_need_vector_tag_u[tag] || _need_vector_tag_dof_u[tag])
+      if ((_sys.subproblem().vectorTagType(tag) == Moose::VECTOR_TAG_RESIDUAL &&
+           _sys.subproblem().safeAccessTaggedVectors()) ||
```
Great. Take both cases

loganharbour committed:
Address reviewer comments

- Uses a map for lookups from TagName to TagID
- Stores the index into the typed vector tags vector in VectorTag

refs #15282:

loganharbour commented:
ping @fdkong 😬

fdkong commented:
> ping @fdkong 😬

I am doing the final check. Sorry
